### PR TITLE
Add the quantized average_pool2d support and adaptive_avg_pool2d support

### PR DESCRIPTION
**Pull request by lly-zero-one**
//copied from PR #25676

===============For avg_pool2d==============

import torch, time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    print('****', str(dtype), '*****')
    x = torch.rand(1, 56, 56, 256)

    q_x = torch.quantize_linear(x, 0.5, 1, dtype)
    q_x = q_x.permute([0, 3, 1, 2])

    x = x.permute([0, 3, 1, 2])

    NITER = 100

    s = time.time()
    for i in range(NITER):
        float_out = torch.nn.functional.avg_pool2d(x, kernel_size=3, stride=None, padding=0)
    time_per_iter_float = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.nn.quantized.functional.avg_pool2d(q_x, kernel_size=3, stride=None, padding=0)
    time_per_iter_quant = (time.time() - s) / NITER

    ref_quantized = torch.quantize_linear(float_out, 0.5, 1, dtype)
    torch.testing.assert_allclose(ref_quantized.dequantize(), quant_out.dequantize())

    print('time/iter ms (float)', 'time/iter ms (quant)', 'quant/float', sep='\t')
    print(time_per_iter_float * 1000, time_per_iter_quant * 1000, time_per_iter_quant / time_per_iter_float, sep='\t')

    bytes_float = (x.numel() + float_out.numel()) * x.element_size()
    bytes_quant = (q_x.numel() + quant_out.numel()) * q_x.element_size()

    float_bw_gbps = bytes_float / time_per_iter_float / 1e9
    quant_bw_gbps = bytes_quant / time_per_iter_quant / 1e9

    print('GB/s float', 'GB/s quant', sep='\t')
    print(float_bw_gbps, quant_bw_gbps, sep='\t')

Before the vectorization:
**** torch.qint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.67439603805542        7.126874923706055       2.6648539791017924
GB/s float      GB/s quant
1.2470733401269298      0.11699265230915809
**** torch.quint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.587001323699951       7.011299133300781       2.7102031487456535
GB/s float      GB/s quant
1.2892022781148076      0.11892118481150399
**** torch.qint32 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.6659250259399414      7.03080415725708        2.637285028215745
GB/s float      GB/s quant
1.2510359321992184      0.4743650833393638

After the vectorization
**** torch.qint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.6113319396972656      0.5631613731384277      0.2156605847679846
GB/s float      GB/s quant
1.2771903676047593      1.48055608884072
**** torch.quint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.5221967697143555      0.5518221855163574      0.21878633425529784
GB/s float      GB/s quant
1.322326647963202       1.5109794819499591
**** torch.qint32 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.5173258781433105      4.0132904052734375      1.5942673295177407
GB/s float      GB/s quant
1.324885279636461       0.8310308159154421

===============For adaptive_avg_pool2d==============
import torch, time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    print('****', str(dtype), '*****')
    x = torch.rand(1, 56, 56, 256)

    q_x = torch.quantize_linear(x, 0.5, 1, dtype)
    q_x = q_x.permute([0, 3, 1, 2])

    x = x.permute([0, 3, 1, 2])

    NITER = 100

    s = time.time()
    for i in range(NITER):
        float_out = torch.nn.functional.adaptive_avg_pool2d(x, output_size=5)
    time_per_iter_float = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.nn.quantized.functional.adaptive_avg_pool2d(q_x, output_size=5)
    time_per_iter_quant = (time.time() - s) / NITER

    ref_quantized = torch.quantize_linear(float_out, 0.5, 1, dtype)
    torch.testing.assert_allclose(ref_quantized.dequantize(), quant_out.dequantize())

    print('time/iter ms (float)', 'time/iter ms (quant)', 'quant/float', sep='\t')
    print(time_per_iter_float * 1000, time_per_iter_quant * 1000, time_per_iter_quant / time_per_iter_float, sep='\t')

    bytes_float = (x.numel() + float_out.numel()) * x.element_size()
    bytes_quant = (q_x.numel() + quant_out.numel()) * q_x.element_size()

    float_bw_gbps = bytes_float / time_per_iter_float / 1e9
    quant_bw_gbps = bytes_quant / time_per_iter_quant / 1e9

    print('GB/s float', 'GB/s quant', sep='\t')
    print(float_bw_gbps, quant_bw_gbps, sep='\t')
~
//Before the vectorization
**** torch.qint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.286238670349121       4.600362777709961       2.0121970804594342
GB/s float      GB/s quant
1.4158031888707898      0.17590264922602994
**** torch.quint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.2867274284362793      4.474163055419922       1.9565790831832832
GB/s float      GB/s quant
1.4155005794518536      0.180864217503144
**** torch.qint32 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.3176145553588867      4.264359474182129       1.8399778618588218
GB/s float      GB/s quant
1.3966360335956578      0.7590504551966285

//After the vectorization:
**** torch.qint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.3224568367004395      0.23195743560791016     0.09987588657942796
GB/s float      GB/s quant
1.3937240722194333      3.4886400510473843
**** torch.quint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.255082130432129       0.2124309539794922      0.09420098324258604
GB/s float      GB/s quant
1.435364129899667       3.8093130254365883
**** torch.qint32 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.266514301300049       1.6029787063598633      0.7072440290539581
GB/s float      GB/s quant
1.4281242338260862      2.0192807222938463

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    scalar_t* output_p,
+    int64_t sizeD,
+    int64_t isizeH,
+    int64_t isizeW,
+    int64_t osizeH,
+    int64_t osizeW,
+    int64_t istrideD,
+    int64_t istrideH,
+    int64_t istrideW) {
+  at::parallel_for(0, sizeD, 0, [&](int64_t start, int64_t end) {
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: Loop order is poor for NHWC, you'll be jumping around reading one byte in multiples of C bytes

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  return output;
+}
+
+Tensor quantized_adaptive_avg_pool2d(
+    const at::Tensor& input,
+    IntArrayRef output_size) {
+  const auto output_shape = get_output_shape(input, output_size);
+  Tensor output = at::_empty_affine_quantized(
+      output_shape, input.options(), input.q_scale(), input.q_zero_point());
+  ;
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: nit: delete

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  adaptive_avg_pool2d_out_template(output, input, output_shape);
+  return output;
+}
+
+Tensor quantized_adaptive_avg_pool2d(
+    const at::Tensor& input,
+    IntArrayRef output_size) {
+  const auto output_shape = get_output_shape(input, output_size);
+  Tensor output = at::_empty_affine_quantized(
+      output_shape, input.options(), input.q_scale(), input.q_zero_point());
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: This does not preserve the layout of the input tensor, meaning you will see poor performance in downstream operators

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  } else {
+    sizeB = input.size(-4);
+    output_shape = {sizeB, sizeD, osizeH, osizeW};
+  }
+
+  return output_shape;
+}
+} // namespace
+
+Tensor& quantized_adaptive_avg_pool2d_out(
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: Is this used?

Additionally, it doesn't seem to me that this takes into account the stride of the output tensor

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <vector>
+
+namespace at {
+namespace native {
+namespace {
+
+inline int start_index(int a, int b, int c) {
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: Please name these function parameters something more descriptive. Same below

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 6]
```diff
@@ -0,0 +1,225 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/Parallel.h>
+#include <ATen/core/op_registration/op_registration.h>
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: Is this include used?

File: aten/src/ATen/native/quantized/cpu/q_avgpool.cpp [hunk 7]
```diff
@@ -12,212 +13,224 @@ namespace at {
...
+          int64_t outputHeight,
+          int kW,
+          int kH,
+          int dW,
+          int dH,
+          int padW,
+          int padH,
+          bool count_include_pad,
+          c10::optional<int64_t> divisor_override)
+{
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/q_avgpool.cpp]: Please clang-format your changes

File: aten/src/ATen/native/quantized/cpu/q_avgpool.cpp [hunk 8]
```diff
@@ -12,212 +13,224 @@ namespace at {
...
-            for (iw = 0; iw < kW; iw++) {
-              int64_t val = (ip + ih * istrideH + iw * istrideW)->val_;
-              sum += val;
+            for(ky = hstart; ky < hend; ky++)
+            {
+              for(kx = wstart; kx < wend; kx++)
+                sum += (ptr_input + ky*inputWidth + kx)->val_;
             }
+            /* Update output */
+            ptr_output->val_ = static_cast<underlying_t>(std::nearbyint(sum/divide_factor));
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/q_avgpool.cpp]: `sum/divide_factor` is done in integer arithmetic. I assume that's not intended?

**Review by jamesr66a — CHANGES_REQUESTED** [review]
Tests are failing due to functional correctness errors.

Additionally, please add a benchmark showing a comparison between this implementation and the float implementation, as well as computed memory bandwidth usage, like in this PR https://github.com/pytorch/pytorch/pull/25676

File: aten/src/ATen/native/quantized/cpu/q_avgpool.cpp [hunk 9]
```diff
@@ -12,212 +13,224 @@ namespace at {
...
+        for(yy = 0; yy < outputHeight; yy++)
+        {
+          for(xx = 0; xx < outputWidth; xx++)
+          {
+            /* Compute the mean of the input image... */
+            int64_t hstart = yy * dH - padH;
+            int64_t wstart = xx * dW - padW;
+            int64_t hend = std::min(hstart + kH, inputHeight + padH);
+            int64_t wend = std::min(wstart + kW, inputWidth + padW);
+            int pool_size = (hend - hstart) * (wend - wstart);
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/q_avgpool.cpp]: Maybe we should use `int64_t` for `pool_size` and `divide_factor`? Or we should change all other variables to be `int`/`int32_t`? Just make it consistent with higher precision (`int64_t`) or the same as C2 (`int32_t`).

**Review by jianyuh — COMMENTED** [review]
This is a C reference implementation. Any plan on also integrating the fast path in https://github.com/pytorch/pytorch/blob/master/caffe2/quantization/server/pool_dnnlowp_op_avx2.cc#L71-L200 ?

**Comment by jamesr66a** [comment]
@jianyuh I saw performance issues with this instruction sequence when I was implementing add https://github.com/pytorch/pytorch/blob/master/caffe2/quantization/server/pool_dnnlowp_op_avx2.cc#L172-L175

Can we instead use a VLEN of 32 and expand out into 4 256-bit float registers? I saw a marked performance improvement when I did that

Btw @llyfacebook  we can also look at my max pool optimized implementation for loop ordering. (It doesn't expand to float but it should be trivial to just slot avg pool in) https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L126

File: test/test_quantized.py [hunk 10]
```diff
@@ -485,9 +485,47 @@ def test_max_pool2d_nhwc(self, X, kernel, stride, dilation, padding):
...
         self.assertEqual(a_ref, a_hat.dequantize(),
                          message="ops.quantized.max_pool2d results are off")
 
+    @given(X=hu.tensor(shapes=hu.array_shapes(min_dims=3, max_dims=4,
+                                              min_side=8, max_side=10),
+                       qparams=hu.qparams()),
+           kernel=st.sampled_from((3, 5)),
+           stride=st.sampled_from((None, 1, 2)),
+           padding=st.integers(0, 2))
+    def test_avg_pool2d(self, X, kernel, stride, padding):
```
Comment (raghuramank100) [inline test/test_quantized.py]: Need to test all fields of the API. The float version looks like this:
```python
F.avg_pool2d(input, self.kernel_size, self.stride,
                            self.padding, self.ceil_mode, self.count_include_pad, self.divisor_override)
```
Specifically, please test for ceil_mode, count_include_pad and divisor override.

File: test/test_quantized.py [hunk 11]
```diff
@@ -485,9 +485,47 @@ def test_max_pool2d_nhwc(self, X, kernel, stride, dilation, padding):
             qa, kernel_size=_pair(kernel),
             stride=_pair(kernel if stride is None else stride),
             padding=_pair(padding), dilation=_pair(dilation))
+        print(a_hat.dequantize())
+        print(a_ref)
         self.assertEqual(a_ref, a_hat.dequantize(),
                          message="ops.quantized.max_pool2d results are off")
 
+    @given(X=hu.tensor(shapes=hu.array_shapes(min_dims=3, max_dims=4,
```
Comment (raghuramank100) [inline test/test_quantized.py]: Please add a test for NHWC layout, similar to what we have for maxpool https://github.com/pytorch/pytorch/pull/25899/files#diff-d7c934f715ae077dce41cdfa38bf0fd1L442

File: aten/src/ATen/native/quantized/cpu/q_avgpool.cpp (hunk 8 above)
Comment (raghuramank100) [reply aten/src/ATen/native/quantized/cpu/q_avgpool.cpp]: That's a good point, The divide should be done in float arithmetic and then rounded to the nearest int, currently both sum and divide_factor are ints.

**Comment by jamesr66a** [comment]
@llyfacebook Please also move the core kernels to QuantizedOpKernels.cpp to ensure they're built with the correct AVX flags

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp (hunk 1 above)
Comment (llyfacebook) [reply aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: Yes, I am implementing a fast path for NHWC layout. This is only for channel-first case.

**Comment by jianyuh** [comment]
> @jianyuh I use a combination of dequantize:
> https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec256/vec256_qint.h#L206
> 
> Operation within loop (constant trip count)(example in add):
> https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L107
> 
> And then requantize:
> https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec256/vec256_qint.h#L76
> 
> Notice the last one doesn't narrow into an int64 but rather just uses a whole vec256 of (u)int8

@jamesr66a : Thanks for the code. I agreed that bundling 4 AVX2 registers together and output a 256-bit of (u)int8 instead of just first 64-bit of (u)int8 can bring more speedup. Will check if we can adopt the same approach in DNNLOWP code once I have more bandwidth.

**Comment by llyfacebook** [comment]
TODO: benchmark data

File: aten/src/ATen/cpu/vec256/vec256_int.h [hunk 12]
```diff
@@ -184,6 +184,12 @@ struct Vec256<int32_t> : public Vec256i {
       std::memcpy(ptr, tmp_values, count * sizeof(int32_t));
     }
   }
+  void dump() const {
+      for (size_t i = 0; i < size(); ++i) {
+          std::cout << (int)((value_type*)&values)[i] << " ";
+      }
+      std::cout << std::endl;
+  }
```
Comment (jamesr66a) [inline aten/src/ATen/cpu/vec256/vec256_int.h]: We should probably just go ahead and add dump() to all Vec256 variants. Can be a separate PR

File: aten/src/ATen/cpu/vec256/vec256_int.h [hunk 13]
```diff
@@ -583,6 +589,18 @@ Vec256<int16_t> inline clamp_min(const Vec256<int16_t>& a, const Vec256<int16_t>
...
+  return _mm256_cvtepi8_epi32(_mm_loadl_epi64(reinterpret_cast<const __m128i*>(ptr)));
+}
+
+Vec256<int32_t> inline convert_8(const int32_t* ptr) {
+  return Vec256<int32_t>::loadu(ptr);
+}
+
+Vec256<int32_t> inline convert_8(const uint8_t* ptr) {
+  return _mm256_cvtepu8_epi32(_mm_loadl_epi64(reinterpret_cast<const __m128i*>(ptr)));
+}
```
Comment (jamesr66a) [inline aten/src/ATen/cpu/vec256/vec256_int.h]: I think a better API would take a size_t template parameter to allow the user to specify how wide they want the conversion. See my discussion about loading full Vec256's of quantized values rather than using `loadl_epi64`. See https://github.com/pytorch/pytorch/blob/b0b0f2c65f4e029e36e17dd8e0f4ebb1c24cb80f/aten/src/ATen/cpu/vec256/vec256_qint.h#L210 where we load a full register and extract and convert subsets of that register

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 14]
```diff
@@ -218,7 +218,187 @@ void qmaxpool_2d_nhwc_kernel(const Tensor &qx,
...
+      int istartH = (int)std::floor((float)(oh * isizeH) / osizeH);
+      int iendH = (int)std::ceil((float)((oh + 1) * isizeH) / osizeH);
+      int kH = iendH - istartH;
+      float kHr = 1.0 / kH;
+      for (int64_t ow = 0; ow < osizeW; ow++) {
+        auto *o_p = reinterpret_cast<typename scalar_t::underlying*>(odata + b * osizeH * osizeW * sizeD + (oh * osizeW + ow) * sizeD);
+        int istartW = (int)std::floor((float)(ow * isizeW) / osizeW);
+        int iendW = (int)std::ceil((float)((ow + 1) * isizeW) / osizeW);;
+        int kW = iendW - istartW;
+        float kHWr = kHr / kW;
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: This variable is unused

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 15]
```diff
@@ -218,7 +218,187 @@ void qmaxpool_2d_nhwc_kernel(const Tensor &qx,
...
+        int64_t c = 0;
+        // For int8 or uint8quantization, we implicitly use int32 as accumulation
+        // Or else, it will go to the slow path
+        // TODO: support 16bit, 32bit, and etc.
+        constexpr auto vec_width = Vec256<scalar_t>::size() / 4;
+        auto* internal_i_p = i_p + istartH * istrideH + istartW * istrideW;
+
+        // TODO: more vectorization with loop interleaving
+#ifdef __AVX2__
+        for (; c + vec_width <= sizeD && vec_width == 8; c += vec_width) {
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Pull `vec_width == 8` into a separate `if` statement around this loop. It doesn't make sense to have it in the loop condition

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 16]
```diff
@@ -218,7 +218,187 @@ void qmaxpool_2d_nhwc_kernel(const Tensor &qx,
...
+          for (int64_t ih = 0; ih < kH; ih++) {
+            for (int64_t iw = 0; iw < kW; iw++) {
+              tcntr = ih * istrideH + iw * istrideW;
+              auto vals = vec256::convert_8(internal_i_p + tcntr + c * istrideD);
+              acc = acc + vals;
+            }
+          }
+          int32_t acc_int[vec_width];
+          float acc_fp[vec_width];
+          // TODO: get rid of the additional copy from vec256<int32> to int32
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: This should get automatically optimized out by the compiler IME

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 17]
```diff
@@ -218,7 +218,187 @@ void qmaxpool_2d_nhwc_kernel(const Tensor &qx,
...
+    scalar_t *idata = static_cast<scalar_t*>(qx.data_ptr());
+    scalar_t *odata = static_cast<scalar_t*>(qy.data_ptr());
+    auto minimum = std::numeric_limits<scalar_t::underlying>::lowest();
+    auto maximum = std::numeric_limits<scalar_t::underlying>::max();
+    int64_t batch_size = nInputPlane*inputWidth*inputHeight;
+    auto *i_p = reinterpret_cast<typename scalar_t::underlying*>(idata + b * batch_size);
+
+    for(int64_t oh = 0; oh < outputHeight; oh++) {
+      for(int64_t ow = 0; ow < outputWidth; ow++) {
+        auto *o_p = reinterpret_cast<typename scalar_t::underlying*>(odata + b * nInputPlane * outputWidth * outputHeight + (oh * outputWidth + ow) * nInputPlane);
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: clang-format your changes, please

File: cmake/Codegen.cmake [hunk 18]
```diff
@@ -115,7 +115,7 @@ if (INTERN_BUILD_ATEN_OPS)
     IF(MSVC)
       LIST(APPEND CPU_CAPABILITY_FLAGS "${OPT_FLAG}/arch:AVX2")
     ELSE(MSVC)
-      LIST(APPEND CPU_CAPABILITY_FLAGS "${OPT_FLAG} -mavx2 -mfma ${CPU_NO_AVX256_SPLIT_FLAGS}")
+      LIST(APPEND CPU_CAPABILITY_FLAGS "${OPT_FLAG} -mavx2 -mavx -mfma ${CPU_NO_AVX256_SPLIT_FLAGS}")
```
Comment (jamesr66a) [inline cmake/Codegen.cmake]: Shouldn't `-mavx` have already been appended from line 98?

**Review by jamesr66a — COMMENTED** [review]
This PR is too long and should be split up. You can probably split it up along the boundary of adding the initial implementation and then adding NHWC specializations and kernels.

BTW are the performance numbers in the description with or without the NHWC specialization?

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+              istrideB,
+              istrideD,
+              istrideH,
+              istrideW);
+        }
+      });
+    }
+    return output;
+  } else {
+    Tensor output = at::_empty_affine_quantized(
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: In this branch `input` might be non contiguous, is it intended or you need to call `input = intput.contiguous()` in this case.

Comment (VitalyFedyunin) [reply aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: Nevermind, see next comment instead.

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+        output_shape,
+        input.options(),
+        input.q_scale(),
+        input.q_zero_point());
+    auto input_contig = input.contiguous();
+    auto input_data = input.data_ptr<scalar_t>();
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: I bet you want `input_contig.data_ptr` here.

File: test/test_quantized.py [hunk 21]
```diff
@@ -488,6 +488,102 @@ def test_max_pool2d_nhwc(self, X, kernel, stride, dilation, padding):
...
+                                              min_side=5, max_side=10),
+                       qparams=hu.qparams()),
+           kernel=st.sampled_from((3, 5)),
+           stride=st.sampled_from((None, 1, 2)),
+           padding=st.integers(0, 2),
+           ceil_mode=st.sampled_from((True, False)),
+           count_include_pad=st.sampled_from((True, False)),
+           divisor_override=st.sampled_from((None, None)))
+    def test_avg_pool2d(self, X, kernel, stride, padding, ceil_mode, count_include_pad, divisor_override):
```
Comment (VitalyFedyunin) [inline test/test_quantized.py]: Please add tests with randomly permuted/strided tensors, might reveal code bugs.

File: aten/src/ATen/native/quantized/cpu/q_avgpool.cpp [hunk 22]
```diff
@@ -12,212 +14,265 @@ namespace at {
...
+                padW,
+                padH,
+                count_include_pad,
+                divisor_override);
+          }
+        });
+      }
+      return output;
+    } else {
+      auto output = at::_empty_affine_quantized(
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/quantized/cpu/q_avgpool.cpp]: Are you expecting input to be contiguous here, or `avg_pool2d_out_frame` can work with any strides?

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp (hunk 20 above)
Comment (llyfacebook) [reply aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: Yes, thanks.

File: aten/src/ATen/cpu/vec256/vec256_int.h (hunk 13 above)
Comment (llyfacebook) [reply aten/src/ATen/cpu/vec256/vec256_int.h]: Yes, loading might be fine. Key is the signed and unsigned. I could use the functional template specialization to make it better.

File: aten/src/ATen/cpu/vec256/vec256_int.h [hunk 23]
```diff
@@ -583,6 +589,21 @@ Vec256<int16_t> inline clamp_min(const Vec256<int16_t>& a, const Vec256<int16_t>
   return _mm256_max_epi16(min_val, a);
 }
 
+template<typename T>
+Vec256<int32_t> inline convert_to_int32(const void* ptr) {
```
Comment (jamesr66a) [inline aten/src/ATen/cpu/vec256/vec256_int.h]: Please type the pointers here and below instead of using `const void*` and doing `reinterpret_cast` immediately on it

File: test/test_quantized.py [hunk 24]
```diff
@@ -531,6 +626,52 @@ def test_adaptive_avg_pool2d(self, X, output_size_h, output_size_w):
                              message=error_message.format(name + '.zero_point', scale,
                                                           qX_hat.q_zero_point()))
 
+    """Tests adaptive average pool operation on NHWC quantized tensors."""
+    @given(X=hu.tensor(shapes=hu.array_shapes(min_dims=4, max_dims=4,
+                                              min_side=1, max_side=10),
+                       qparams=hu.qparams(dtypes=torch.qint8)),
```
Comment (jamesr66a) [inline test/test_quantized.py]: This is only being tested on `torch.qint8` because of this line. Can you omit the `dtypes` kwarg?

**Review by jamesr66a — CHANGES_REQUESTED** [review]
Performance results are 25x slower than expected. This indicates either the vectorized implementation is wrong or it is not being run

**Review by jamesr66a — APPROVED** [review]
Discussed offline. The slowdown is likely because we consistently hit the scalar code in the QuantizeAVX2 routine: https://github.com/pytorch/pytorch/blob/0f42881269d10056108e51320533ea2e55ade221/aten/src/ATen/cpu/vec256/vec256_qint.h#L124

I'm okay with moving the performance optimization to a follow-up PR since this one is already gigantic

Please also fix up my last comments before merging
